zcash_primitives/src/transaction/mod.rs

```diff
@@ -125,12 +126,26 @@ impl TransactionData {
         header
     }
 
-    pub fn freeze(self) -> Transaction {
-        Transaction(self)
+    pub fn freeze(self) -> io::Result<Transaction> {
+        Transaction::from_data(self)
     }
 }
 
 impl Transaction {
+    fn from_data(data: TransactionData) -> io::Result<Self> {
+        let mut tx = Transaction(data, TxId([0; 32]));
+        let mut raw = vec![];
+        tx.write(&mut raw)?;
+        (tx.1)
+            .0
+            .copy_from_slice(&Sha256::digest(&Sha256::digest(&raw)));
+        Ok(tx)
+    }
+
+    pub fn txid(&self) -> TxId {
+        self.1
```

Comment:
I am not sure that this method is used anywhere. If txid were a named field, would this method still be necessary?

Reply:
The function isn't used anywhere because nothing in the librustzcash ecosystem yet requires it, but downstream consumers definitely will.

Making this a public named field would enable anyone with a mutable `Transaction` reference to modify the `txid`. We protect the rest of the transaction properties by having them be part of `TransactionData`, but we can't do that for `txid` because we can't fix the transaction ID until we freeze the transaction. Thus IMHO a getter makes more sense in this context.

FYI, a getter with the same name as an underlying field is [consistent with Rust naming conventions](https://doc.rust-lang.org/1.0.0/style/style/naming/README.html#getter/setter-methods-[rfc-344]).